<a href="https://colab.research.google.com/github/SanketIngole19/Glassdoor-Salary-Prediction/blob/main/Glassdoor_Salary_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Glassdoor Salary Prediction

Github Link - https://github.com/SanketIngole19/Glassdoor-Salary-Prediction

Project Type - EDA/LinearRegression/NLP/RandomForestRegression/Stacking Regressor
Contribution - Individual

Project Summary -
In the fast-paced tech industry, salary trends are influenced by factors like job role, company size, experience, and location. Understanding these patterns is essential for job seekers, employers, and policymakers.

This project analyzes Glassdoor job postings data (2017) to predict salaries for various tech roles. By examining key features such as job title, company size, and location, it uncovers salary trends, compares compensation across industries, and provides valuable insights for salary expectations

Problem Statement -
How does salary vary by job position (e.g., Data Scientist vs. Software Engineer vs. DevOps Engineer)?

What is the impact of company size on salary levels? How do salaries differ by location (e.g., San Francisco vs. Austin vs. New York)?

Can we build a predictive model to estimate salaries based on job attributes?


Import Libraries

In [ ]:
# Import Libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


Dataset Loading

In [ ]:
# Load Dataset
df = pd.read_csv('/content/drive/MyDrive/glassdoor_jobs.csv')


Dataset First View

In [ ]:
# Dataset First Look
df.head(2)

Unnamed: 0                  Job Title              Salary Estimate  \
0           0             Data Scientist   $53K-$91K (Glassdoor est.)   
1           1  Healthcare Data Scientist  $63K-$112K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Data Scientist\nLocation: Albuquerque, NM\nEdu...     3.8   
1  What You Will Do:\n\nI. General Summary\n\nThe...     3.4   

                                 Company Name         Location   Headquarters  \
0                      Tecolote Research\n3.8  Albuquerque, NM     Goleta, CA   
1  University of Maryland Medical System\n3.4    Linthicum, MD  Baltimore, MD   

                    Size  Founded   Type of ownership  \
0  501 to 1000 employees     1973   Company - Private   
1       10000+ employees     1984  Other Organization   

                           Industry               Sector  \
0               Aerospace & Defense  Aerospace & Defense   
1  Health Care Services & Hospitals          Health Care   

                     Revenue Competitors  
0  $50 to $100 million (USD)          -1  
1     $2 to $5 billion (USD)          -1

Dataset Rows & Columns count

In [ ]:
# Dataset Rows & Columns count
df.shape

(956, 15)

Dataset Information

In [ ]:
# Dataset Info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 956 entries, 0 to 955
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         956 non-null    int64  
 1   Job Title          956 non-null    object 
 2   Salary Estimate    956 non-null    object 
 3   Job Description    956 non-null    object 
 4   Rating             956 non-null    float64
 5   Company Name       956 non-null    object 
 6   Location           956 non-null    object 
 7   Headquarters       956 non-null    object 
 8   Size               956 non-null    object 
 9   Founded            956 non-null    int64  
 10  Type of ownership  956 non-null    object 
 11  Industry           956 non-null    object 
 12  Sector             956 non-null    object 
 13  Revenue            956 non-null    object 
 14  Competitors        956 non-null    object 
dtypes: float64(1), int64(2), object(12)
memory usage: 112.2+ KB


Duplicate Values

In [ ]:
# Dataset Duplicate Value Count
df.duplicated().sum()

np.int64(0)

Missing Values/Null Values

In [ ]:
# Missing Values/Null Values Count
df.isnull().sum()

Unnamed: 0           0
Job Title            0
Salary Estimate      0
Job Description      0
Rating               0
Company Name         0
Location             0
Headquarters         0
Size                 0
Founded              0
Type of ownership    0
Industry             0
Sector               0
Revenue              0
Competitors          0
dtype: int64

What did you know about your dataset?

The Dataset has Jobtitles and its relevant informations like Job description, salary range, job location and as such.

The Dataset has total 956 rows and 15 columns

There are no null/missing values in the dataset

There are no duplicate entries in the dataset

. Understanding Your Variables

In [ ]:
# Dataset Columns
df.columns

Index(['Unnamed: 0', 'Job Title', 'Salary Estimate', 'Job Description',
       'Rating', 'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors'],
      dtype='object')

In [ ]:
# Dataset Describe
df.describe()

Unnamed: 0      Rating      Founded
count  956.000000  956.000000   956.000000
mean   477.500000    3.601255  1774.605649
std    276.117729    1.067619   598.942517
min      0.000000   -1.000000    -1.000000
25%    238.750000    3.300000  1937.000000
50%    477.500000    3.800000  1992.000000
75%    716.250000    4.200000  2008.000000
max    955.000000    5.000000  2019.000000

Check Unique Values for each variable.

In [ ]:
# Check Unique Values for each variable.
variables = ['Job Title', 'Company Name', 'Location', 'Headquarters',
       'Type of ownership', 'Industry', 'Sector', 'Revenue',]

for variable in variables:
  print(f'Unique variables in {variable} are {df[variable].unique()}','\n')

Unique variables in Job Title are ['Data Scientist' 'Healthcare Data Scientist' 'Research Scientist'
 'Staff Data Scientist - Technology' 'Data Analyst' 'Data Engineer I'
 'Scientist I/II, Biology' 'Customer Data Scientist'
 'Data Scientist - Health Data Analytics'
 'Senior Data Scientist / Machine Learning'
 'Data Scientist - Quantitative' 'Digital Health Data Scientist'
 'Associate Data Analyst' 'Clinical Data Scientist'
 'Data Scientist / Machine Learning Expert' 'Web Data Analyst'
 'Senior Data Scientist' 'Data Engineer'
 'Data Scientist - Algorithms & Inference' 'Scientist'
 'Data Science Analyst' 'Lead Data Scientist'
 'Spectral Scientist/Engineer'
 'College Hire - Data Scientist - Open to December 2019 Graduates'
 'Data Scientist, Office of Data Science' 'Business Intelligence Analyst'
 'Senior Risk Data Scientist'
 'Data Scientist in Artificial Intelligence Early Career'
 'Data Scientist - Research' 'R&D Data Analysis Scientist'
 'Analytics Consultant' 'Director, Data Science' 

Data Wrangling

Data Wrangling Code

In [ ]:
# dropping the Unnamed: 0 column, as it is of no significance

df.drop('Unnamed: 0',axis=1,inplace=True)
df.columns

Index(['Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors'],
      dtype='object')

In [ ]:
# -1 in competitors are missing values, hence substituting the values with NA
df['Competitors'] = df['Competitors'].replace('-1',pd.NA)


In [ ]:
df[['Minimum Salary', 'Maximum Salary']] = df['Salary Estimate'].str.replace(r'\s*\(Glassdoor est.\)', '', regex=True).str.replace("$", "").str.replace("K", "000").str.split("-", expand=True)


In [ ]:
df['Maximum Salary'] = df['Maximum Salary'].str.replace(r'\(Employer est.\)', '', regex = True)
df['Minimum Salary'] = df['Minimum Salary'].str.replace(r'\(Employer est.\)', '', regex = True)

In [ ]:
# Remove non-numeric characters
df['Maximum Salary'] = df['Maximum Salary'].str.extract(r'(\d+\.?\d*)')[0].astype(float)
df['Minimum Salary'] = df['Minimum Salary'].str.extract(r'(\d+\.?\d*)')[0].astype(float)

In [ ]:
df['Average Salary'] = (df['Maximum Salary'] + df['Minimum Salary'])/2

In [ ]:
df[['Company Name clean','Company Rating']] = df['Company Name'].str.split('\n', expand=True)
df['Company Rating'] = df['Company Rating'].astype(float)

In [ ]:
#dropped company rating as a seprate column for Rating already exists

df.drop('Company Rating', inplace = True, axis =1)

In [ ]:
#Replacing all -1 values from Founded to nan
df['Founded'] = df['Founded'].replace(-1,np.nan)

In [ ]:
# calculating company age

df['Company age'] = 2025 - df['Founded']

In [ ]:
#extracting minimum and maximum number of employees

df[['Min_emp','Max_emp']] = df['Size'].str.split(' to ', expand =True)
df['Min_emp'] = pd.to_numeric(df['Min_emp'].str.replace(r'\D','',regex=True), errors='coerce').astype('Int64')

df['Max_emp'] = pd.to_numeric(df['Max_emp'].fillna(0).str.split(' ').str[0], errors='coerce').fillna(0).astype(int)

In [ ]:
df['Average_emp'] = abs((df['Max_emp'] + df['Min_emp'])/2)

In [ ]:
Cleaned_df = df.copy()

In [ ]:
Cleaned_df.columns

Index(['Job Title', 'Salary Estimate', 'Job Description', 'Rating',
       'Company Name', 'Location', 'Headquarters', 'Size', 'Founded',
       'Type of ownership', 'Industry', 'Sector', 'Revenue', 'Competitors',
       'Minimum Salary', 'Maximum Salary', 'Average Salary',
       'Company Name clean', 'Company age', 'Min_emp', 'Max_emp',
       'Average_emp'],
      dtype='object')

In [ ]:
#Company Paying the maximum salary

Cleaned_df[Cleaned_df['Maximum Salary'] == max(Cleaned_df['Maximum Salary'])]

Job Title               Salary Estimate  \
452  Director II, Data Science - GRM Actuarial  $202K-$306K (Glassdoor est.)   
675  Director II, Data Science - GRM Actuarial  $202K-$306K (Glassdoor est.)   
913  Director II, Data Science - GRM Actuarial  $202K-$306K (Glassdoor est.)   

                                       Job Description  Rating  \
452  Advance your career at Liberty Mutual Insuranc...     3.3   
675  Advance your career at Liberty Mutual Insuranc...     3.3   
913  Advance your career at Liberty Mutual Insuranc...     3.3   

                      Company Name     Location Headquarters  \
452  Liberty Mutual Insurance\n3.3  Chicago, IL   Boston, MA   
675  Liberty Mutual Insurance\n3.3  Chicago, IL   Boston, MA   
913  Liberty Mutual Insurance\n3.3  Chicago, IL   Boston, MA   

                 Size  Founded  Type of ownership  ...             Revenue  \
452  10000+ employees   1912.0  Company - Private  ...  $10+ billion (USD)   
675  10000+ employees   1912.0  Company - Private  ...  $10+ billion (USD)   
913  10000+ employees   1912.0  Company - Private  ...  $10+ billion (USD)   

                         Competitors Minimum Salary Maximum Salary  \
452  Travelers, Allstate, State Farm       202000.0       306000.0   
675  Travelers, Allstate, State Farm       202000.0       306000.0   
913  Travelers, Allstate, State Farm       202000.0       306000.0   

     Average Salary        Company Name clean  Company age Min_emp  Max_emp  \
452        254000.0  Liberty Mutual Insurance        113.0   10000        0   
675        254000.0  Liberty Mutual Insurance        113.0   10000        0   
913        254000.0  Liberty Mutual Insurance        113.0   10000        0   

     Average_emp  
452       5000.0  
675       5000.0  
913       5000.0  

[3 rows x 22 columns]

In [ ]:
#Companies paying the least salary
Cleaned_df[Cleaned_df['Minimum Salary'] == min(Cleaned_df['Minimum Salary'])]

Job Title  \
523  Senior Operations Data Analyst, Call Center Op...   
787  Senior Operations Data Analyst, Call Center Op...   

                      Salary Estimate  \
523  $10-$17 Per Hour(Glassdoor est.)   
787  $10-$17 Per Hour(Glassdoor est.)   

                                       Job Description  Rating   Company Name  \
523  Summary\n\n\nIndependent contributor role that...     2.7  FLEETCOR\n2.7   
787  Summary\n\n\nIndependent contributor role that...     2.7  FLEETCOR\n2.7   

          Location           Headquarters                     Size  Founded  \
523  Nashville, TN  Peachtree Corners, GA  5001 to 10000 employees   2000.0   
787  Nashville, TN  Peachtree Corners, GA  5001 to 10000 employees   2000.0   

    Type of ownership  ...                 Revenue Competitors Minimum Salary  \
523  Company - Public  ...  $2 to $5 billion (USD)        <NA>           10.0   
787  Company - Public  ...  $2 to $5 billion (USD)        <NA>           10.0   

    Maximum Salary  Average Salary  Company Name clean  Company age Min_emp  \
523           17.0            13.5            FLEETCOR         25.0    5001   
787           17.0            13.5            FLEETCOR         25.0    5001   

     Max_emp  Average_emp  
523    10000       7500.5  
787    10000       7500.5  

[2 rows x 22 columns]

In [ ]:
#Highest Rated Companies

Cleaned_df[Cleaned_df['Rating']==5.0]['Company Name clean'].unique()

array(['Brightside', 'BPA Services', 'Alector', 'Gridiron IT',
       'Hamilton Porter LLC', 'Aptonet Inc', 'Genesis Research', '605',
       'NS8', 'Freedman Healthcare', 'Senti Biosciences',
       'Royce Geospatial', 'Quality Consulting Group',
       'Edison Software, Inc.', 'SkySync',
       'HireResources a Tailored Solutions Company (WO)', 'Jobot'],
      dtype=object)

In [ ]:
#Oldest company
Cleaned_df[Cleaned_df['Company age']== max(Cleaned_df['Company age'])]

Job Title              Salary Estimate  \
347  Data Scientist  $95K-$161K (Glassdoor est.)   

                                       Job Description  Rating  \
347  Overview\n\n\nAt Sotheby’s we use data science...     3.6   

       Company Name      Location  Headquarters                    Size  \
347  Sotheby's\n3.6  New York, NY  New York, NY  1001 to 5000 employees   

     Founded Type of ownership  ...                           Revenue  \
347   1744.0  Company - Public  ...  $500 million to $1 billion (USD)   

    Competitors Minimum Salary Maximum Salary  Average Salary  \
347        <NA>        95000.0       161000.0        128000.0   

     Company Name clean  Company age Min_emp  Max_emp  Average_emp  
347           Sotheby's        281.0    1001     5000       3000.5  

[1 rows x 22 columns]

In [ ]:
#Youngest Company
Cleaned_df[Cleaned_df['Company age']== min(Cleaned_df['Company age'])]['Company Name clean'].unique()

array(['Sapphire Digital', 'Hamilton Porter LLC', 'U.Group'], dtype=object)

In [ ]:
#Highest number of employess
Cleaned_df[Cleaned_df['Max_emp'] == max(Cleaned_df['Max_emp'])]['Company Name clean'].unique()

array(['ManTech', 'The Hanover Insurance Group', 'Juniper Networks',
       'Sartorius', 'Edgewell Personal Care', 'Intrado',
       'Johns Hopkins University Applied Physics Laboratory',
       'TransUnion', 'MITRE', 'Autodesk', 'Insight Enterprises, Inc.',
       'BlueCross BlueShield of Tennessee', 'MassMutual', 'Tower Health',
       'BWX Technologies', 'Reynolds American', 'Genesys',
       'AmeriHealth Caritas', 'FLEETCOR', 'Mentor Graphics',
       'Maxar Technologies'], dtype=object)

In [ ]:
Cleaned_df['Size'] = Cleaned_df['Size'].replace('-1',np.nan)
Cleaned_df['Size'] = Cleaned_df['Size'].replace('Unknown',np.nan)

Hypothesis Testing

Based on your chart experiments, define three hypothetical statements from the dataset. In the next three questions, perform hypothesis testing to obtain final conclusion about the statements through your code and statistical testing.

Hypothetical Statement - 1

Companies with higher ratings tend to offer higher salaries.

Null Hypothesis (H₀): No significant difference in salaries between high-rated and low-rated companies.

Alternative Hypothesis (H₁): Higher-rated companies (≥4.0) offer significantly higher salaries.

Hypothetical Statement - 2

Older Companies Pay More

Null Hypothesis (H₀): No significant difference in salaries between age of company

Alternative Hypothesis (H₁): Older companies pay more

Hypothetical Statement - 3

Older Companies are rated high

Null Hypothesis (H₀): There is no significant difference in company ratings between older and newer companies.

Alternative Hypothesis (H₁): Older companies have significantly higher ratings than newer companies.

Feature Engineering & Data Pre-processing

Handling Missing Values

In [ ]:
# Handling Missing Values & Missing Value Imputation

Cleaned_df.isnull().sum()[Cleaned_df.isnull().sum()>0]

Size               26
Founded            97
Competitors       634
Minimum Salary    214
Average Salary    214
Company age        97
Min_emp            15
Average_emp        15
dtype: int64

In [ ]:
from sklearn.impute import SimpleImputer

median_imputer = SimpleImputer(strategy ='median')
mean_imputer = SimpleImputer(strategy='mean')

#founded is not normally distriubted, hence using median imputation
Cleaned_df['Founded'] = median_imputer.fit_transform(Cleaned_df[['Founded']])

Cleaned_df.drop('Competitors', axis = 1,inplace = True)

#minimum salary is normally distributed, hence using mean imputation
Cleaned_df['Minimum Salary'] = mean_imputer.fit_transform(Cleaned_df[['Minimum Salary']])

#since minimum salary's missing values are handeled, we are recalculating the average salary to fill in the missing values
Cleaned_df['Average Salary'] = (Cleaned_df['Maximum Salary'] + Cleaned_df['Minimum Salary'])/2

#founded missing values have been handeled, recalculating the company age to fill in missing values
Cleaned_df['Company age'] = 2025 - Cleaned_df['Founded']

#replacing the missing values in minimum number of employees with median values as the distribution is not normal
Cleaned_df['Min_emp'] = median_imputer.fit_transform(Cleaned_df[['Min_emp']])

#recalculating avg emp column as the minimum emp is recalculated
Cleaned_df['Average_emp'] = (Cleaned_df['Max_emp'] + Cleaned_df['Min_emp'])/2

#Replaced Unknow and -1 from Type of Ownership with Other Organization
Cleaned_df['Type of ownership'].replace(['Unknown','-1'],'Other Organization',inplace = True)

#Some of the companies ratings are -1 which is not possible, these may be defalut values filling in for missing values
Cleaned_df['Rating'].replace(-1, round(Cleaned_df['Rating'].mean(),1), inplace=True)

/tmp/ipython-input-37-4247733703.py:27: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  Cleaned_df['Type of ownership'].replace(['Unknown','-1'],'Other Organization',inplace = True)
/tmp/ipython-input-37-4247733703.py:30: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, whe

In [ ]:
Cleaned_df.isnull().sum()[Cleaned_df.isnull().sum()>0]

Size    26
dtype: int64

In [ ]:
Cleaned_df['Founded'] = Cleaned_df['Founded'].astype(int)

In [ ]:
# Encode your categorical columns

Cleaned_df['Sector'] = Cleaned_df['Sector'].astype("category")
Cleaned_df['Type of ownership'] = Cleaned_df['Type of ownership'].astype("category")

In [ ]:
!pip install category_encoders

import category_encoders as ce
label_encoder = ce.OrdinalEncoder()
Cleaned_df['encoded_sector'] = label_encoder.fit_transform(Cleaned_df['Sector'])
Cleaned_df['encoded_type_of_ownership'] = label_encoder.fit_transform(Cleaned_df['Type of ownership'])

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.7/85.7 kB 2.4 MB/s eta 0:00:00


Conclusion

This project successfully analyzed salary trends in the tech industry using a 2017 Glassdoor dataset, addressing key questions about salary variations by job position, company size, and location.

Through extensive exploratory data analysis (EDA), 15 insightful visualizations revealed patterns such as salary disparities, the prevalence of mid-sized companies, and high-demand job locations.

Feature engineering, including TF-IDF vectorization of job descriptions and categorical encoding, enabled robust model development.

Three machine learning models—Linear Regression, Random Forest Regressor, and Stacking Regressor—were implemented, with the Stacking Regressor (combining XGBoost, Bayesian Ridge, and SVR) achieving the best performance after cross-validation and hyperparameter tuning **(R²: 0.97, RMSE:  0.17)**.

SHAP analysis highlighted encoded_job_title and encoded_company_name as the most influential features for salary prediction.

The final model was saved for deployment, providing a reliable tool for estimating salaries based on job attributes, which can benefit job seekers, employers, and policymakers in understanding and navigating compensation trends.




### ***Hurrah! You have successfully completed your Machine Learning Capstone Project !!!***